### PR TITLE
Add ``ScalarOp`` to `quantum_info.operators`

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Adds two new operator objects:
* `ScalarOp` which can represent a  multi-partite identity operator with an optional complex scale coefficient.
* ~`ZeroOp` which is a multi-partite zero matrix operator.~

Depends on #3873 

### Details and comments

These operators can be added / composed with other operators and will handle automatic conversion to the other operator types. And are designed for initializing a trivial system and to be used with other BaseOperator subclasses.

The ScalarOp can used for composition or addition. Eg initializing a multi-system identity and composing other operators/gates with it on subsystems.

~The ZeroOp can be used for addition/subtraction. Eg: initializing a multi-system zero matrix and adding components on subsystems (this requires a future PR to implement #3877 which will depend on this PR)~ *ZeroOp is removed and can instead be done using `ScalarOp(dims, coeff=0)`*.

For example the following will return an `Operator` object for `XI`.

```python
iden = ScalarOp([2, 2, 2])  # 3 qubit identity
X = Operator.from_label('X')
iden @ X([1])
```

For example the following will return an `SuperOp` object for unitary channel `XI`.

```python
iden = ScalarOp([2, 2, 2])  # 3 qubit identity
X = SuperOp(Operator.from_label('X'))
iden @ X([1])
```

### CODE REVIEW

File: qiskit/quantum_info/operators/identity_op.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+from numbers import Number
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.quantum_info.operators.operator import Operator
+
+
+class IdentityOp(BaseOperator):
```
Comment (ewinston) [inline qiskit/quantum_info/operators/identity_op.py]: Would it be more accurate to call this `ScalarOp` since it takes a coefficient?

File: test/python/quantum_info/operators/test_identity_op.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+                    elif coeff2 is None:
+                        target = coeff1
+                    else:
+                        target = coeff1 * coeff2
+                    self.assertTrue(isinstance(val, IdentityOp))
+                    self.assertEqual(val.input_dims(), dims)
+                    self.assertEqual(val.output_dims(), dims)
+                    self.assertAlmostEqual(val.coeff, target)
+
+    def test_compose_qargs_identity(self):
```
Comment (ewinston) [inline test/python/quantum_info/operators/test_identity_op.py]: Would larger qargs, e.g. `qargs=[0, 2]`, also work?

File: qiskit/quantum_info/operators/identity_op.py (hunk 1 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/identity_op.py]: Thats a possibility. Or make a second op called ScalarOp and have identity not have a coeff.

File: test/python/quantum_info/operators/test_identity_op.py (hunk 2 above)
Comment (chriseclectic) [reply test/python/quantum_info/operators/test_identity_op.py]: Yep any qargs work if there are enough subsystems.

File: qiskit/quantum_info/operators/scalar_op.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    """Scalar identity operator class.
+
+    This is a symbolic representation of an scalar identity operator on
+    multiple subsystems. It may be used to initialize a symbolic scalar
+    multiplication of an identity and then be implicitly converted to other
+    kinds of operator subclasses by using the :meth:`compose`, :meth:`dot`,
+    :meth:`tensor`, :meth:`expand` methods.
+    """
+
+    def __init__(self, dims, coeff=None):
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]:
```suggestion
    def __init__(self, dims, coeff=1):
```
If you default to 1 here you save a lot of checking for None in the code. Everytime you check for None that's basically a way of saying coefficient is 1.

File: qiskit/quantum_info/operators/scalar_op.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+from numbers import Number
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.quantum_info.operators.operator import Operator
+
+
+class ScalarOp(BaseOperator):
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: shouldn't this just derive from Operator?

File: qiskit/quantum_info/operators/scalar_op.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        """Convert to an Operator object."""
+        return Operator(self.to_matrix(),
+                        input_dims=self._input_dims,
+                        output_dims=self._output_dims)
+
+    def tensor(self, other):
+        """Return the tensor product operator self ⊗ other.
+
+        Args:
+            other (BaseOperator): an operator object.
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: can you tensor a ScalarOp with a QuantumChannel? If not this should just be Operator I think (would help if ScalarOp was derived from Operator like I commented earlier)

File: qiskit/quantum_info/operators/scalar_op.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                coeff = self.coeff * other.coeff
+            dims = other._input_dims + self._input_dims
+            return ScalarOp(dims, coeff=coeff)
+        return other.expand(self)
+
+    def expand(self, other):
+        """Return the tensor product operator other ⊗ self.
+
+        Args:
+            other (ScalarOp or Operator): an operator object.
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: this should have the same arg type as tensor().

File: qiskit/quantum_info/operators/scalar_op.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            else:
+                coeff = self.coeff * other.coeff
+            return ScalarOp(input_dims, coeff=coeff)
+
+        # If we are composing on the full system we return the
+        # other operator with reshaped dimensions
+        if qargs is None:
+            ret = other.reshape(input_dims, output_dims)
+            if self.coeff is None or self.coeff == 1:
+                return ret
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: see my previous comment about the None default which can just be 1.

File: qiskit/quantum_info/operators/scalar_op.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+        Raises:
+            QiskitError: if the input and output dimensions of the operator
+            are not equal, or the power is not a positive integer.
+        """
+        if self.coeff is None:
+            return self
+        if n == 0:
+            # Raising to zero power returns identity
+            return ScalarOp(self._input_dims)
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: even if you remove this block it still does this (only difference is maybe you get `coeff=1+0j` rather than `coeff=1`, which I basically the same). In general I think there are too many manual `if` checking, whereas math will just make things consistent.

File: qiskit/quantum_info/operators/scalar_op.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            coeff1 = 1 if self.coeff is None else self.coeff
+            coeff2 = 1 if other.coeff is None else other.coeff
+            return ScalarOp(self._input_dims, coeff=coeff1+coeff2)
+        return other._add(self).reshape(self._input_dims, self._output_dims)
+
+    def _multiply(self, other):
+        """Return the ScalarOp other * self.
+
+        Args:
+            other (complex): a complex number.
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]:
```suggestion
            other (Number): a complex number.
```

File: qiskit/quantum_info/operators/scalar_op.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+        Returns:
+            ScalarOp: the scaled identity operator other * self.
+
+        Raises:
+            QiskitError: if other is not a valid complex number.
+        """
+        if not isinstance(other, Number):
+            raise QiskitError("other is not a number")
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: this is exactly the type of thing type hints do, so you don't have to riddle your code with manual type checks

File: qiskit/quantum_info/operators/scalar_op.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+            ret = other.reshape(input_dims, output_dims)
+            if self.coeff is None or self.coeff == 1:
+                return ret
+            return self.coeff * ret
+        # Otherwise compose using other operators method
+        # Note that in this case that operator must know how to initalize
+        # from an ScalarOp either using its to_operator method or having
+        # a specific initialization case.
+        return other.__class__(self).compose(
+            other, qargs=qargs, front=not front)
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/scalar_op.py]: can you explain what's going on here? if no `qargs` are specified during compose, you compose on the full subsystem.. so qargs would become equal to `other.qargs` before reaching this point. would that work if other is smaller or larger than self?

File: test/python/quantum_info/operators/test_scalar_op.py [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+
+    def test_compose_operator(self):
+        """Test compose and dot methods with ScalarOp and Operator."""
+        for coeff in [None, 1, -3.1, 1 + 3j]:
+            for label in ['II', 'XX', 'YY', 'ZZ']:
+                dim = 4
+                iden = ScalarOp(dim, coeff=coeff)
+                op = Operator.from_label(label)
+
+                with self.subTest(msg='{}.compose(Operator({}))'.format(iden, label)):
```
Comment (ajavadia) [inline test/python/quantum_info/operators/test_scalar_op.py]: @1ucian0 can something like this use the `combine` decorator of tests, instead of using subTest? or is subTest fine?

File: qiskit/quantum_info/operators/scalar_op.py (hunk 11 above)
Comment (ajavadia) [reply qiskit/quantum_info/operators/scalar_op.py]: i see erick asked a similar question, but i just couldn't follow from the code.

File: qiskit/quantum_info/operators/scalar_op.py (hunk 4 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: No, then it would inherit its array fields and all the other unnecessary numerical baggage.

File: qiskit/quantum_info/operators/scalar_op.py (hunk 5 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: Yes you can tensor with QuantumChannel. the idea is you can tensor with anything derived from BaseOperator

File: qiskit/quantum_info/operators/scalar_op.py (hunk 10 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: I disagree. As far as I can tell Type hints are just hints, they don't stop you inserting any value and don't do error checking. eg: the following will execute without any warnings or errors

```
def foo(val: str) -> str:
    return val

foo([1,2,3])
```

File: qiskit/quantum_info/operators/scalar_op.py (hunk 3 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: This could be changed

File: qiskit/quantum_info/operators/scalar_op.py (hunk 11 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: This allows the constructor for another BaseOperator subclass to handle conversion of the identity (or scalar identity) to whatever data format it has, and then composes the other operator with it.

Eg if it was a matrix it would convert to the identity matrix and then compose.
If it was a Choi it would convert to identity channel Choi matrix then compose
If it was a Clifford it would convert to identityt Clifford then compose etc...

Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: I think the `front= not front` is a typo, not that order makes a difference for scalar op.

File: qiskit/quantum_info/operators/scalar_op.py (hunk 8 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/scalar_op.py]: This was just being pedantic, so could be removed

File: test/python/quantum_info/operators/test_scalar_op.py (hunk 12 above)
Comment (1ucian0) [reply test/python/quantum_info/operators/test_scalar_op.py]: I would say both, at least in some scenarios.  I had trouble sleeping so I refactor the tests a bit in 73d5cc5eb. I think they are much easier to read now.

File: test/python/quantum_info/operators/test_zero_op.py [hunk 13]
```diff
@@ -0,0 +1,506 @@
...
+"""Tests for ZeroOp class."""
+
+import unittest
+import numpy as np
+
+from qiskit.test import QiskitTestCase
+from qiskit.quantum_info.operators import ZeroOp, ScalarOp, Operator
+
+
+class TestZeroOp(QiskitTestCase):
```
Comment (1ucian0) [inline test/python/quantum_info/operators/test_zero_op.py]: These tests, as test/python/quantum_info/operators/test_scalar_op.py before 73d5cc5, seem too monolithic and the could be broken. Breaking tests in smaller cases give more information about fails when they fail.

File: test/python/quantum_info/operators/test_scalar_op.py (hunk 12 above)
Comment (chriseclectic) [reply test/python/quantum_info/operators/test_scalar_op.py]: I like using subTest because it shows each failing subtest as a separate test failure when something goes wrong.

File: test/python/quantum_info/operators/test_zero_op.py (hunk 13 above)
Comment (chriseclectic) [reply test/python/quantum_info/operators/test_zero_op.py]: I thought that was the whole point of subTests: each failing subtest shows as separate failure when they fail.

**Comment by chriseclectic** [comment]
@ajavadia After changing `IdentityOp` to `ScalarOp`, I don't really see the need for a separate `ZeroOp`, it can just be `ScalarOp(dims, coeff=0)` and the utility for that class can handled as a special case of the `_add` method when `coeff=0`.

The only loss here is that you can't have a generic rectangular ZeroOp (ScalarOp is always square), but this is a not often used edge case so I don't think it matters to much.

File: test/python/quantum_info/operators/test_scalar_op.py (hunk 12 above)
Comment (1ucian0) [reply test/python/quantum_info/operators/test_scalar_op.py]: `combine` also allows that and, in addition, you can run the failing test independently of the rest of the cases. So your breakpoints for debugging are easier set.

File: test/python/quantum_info/operators/test_zero_op.py (hunk 13 above)
Comment (1ucian0) [reply test/python/quantum_info/operators/test_zero_op.py]: But subTests cannot run independently. Also, the subTests make easier to accidentally create dependency between tests, which is something not recommended...

Sometimes unfolding a subtest is too much work and does not increase the case coverage much (somebody can argue that that test is necessary anyway)... so I'm not saying that is totally removable in every situation.

Comment (chriseclectic) [reply test/python/quantum_info/operators/test_zero_op.py]: Fair enough

File: qiskit/quantum_info/operators/scalar_op.py (hunk 10 above)
Comment (ajavadia) [reply qiskit/quantum_info/operators/scalar_op.py]: yes, python still remains a dynamically-typed language. However if the code has type hints other tools (such as MyPy) can catch programming errors when you use a disallowed type, without running the code. 

But my point here was you already say the expected type (in docstring in this case), so checking for it is a bit redundant unless there is a real pitfall of programmers passing in the wrong type.

**Comment by chriseclectic** [comment]
@ajavadia I added 0 to all the test coeffs lists as a test for that case in all the methods.
